### PR TITLE
add must use annotation to functions that have no mutable args or statics

### LINKED ISSUE #57: Fix clippy::must-use-candidate warnings
I ran Clippy's pedantic lints (which are disabled by default since they are typically subjective or have high false positive rates) against this repository out of curiosity, and am filing a few issues for some of the lints that warned that seemed relevant (in part to give some easy issues for people to work through to get involved with this library).

Feel free to close this out if not wanted :-)

See:
https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

```
$ cargo clippy --all-targets --all-features -- -W clippy::must-use-candidate
    Checking libcnb v0.2.0 (/Users/emorley/src/libcnb.rs)
warning: this method could have a `#[must_use]` attribute
  --> src/data/build_plan.rs:16:5
   |
16 |     pub fn new() -> BuildPlan {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn new() -> BuildPlan`
   |
   = note: requested on the command line with `-W clippy::must-use-candidate`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/data/build_plan.rs:38:5
   |
38 |     pub fn new() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn new() -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/data/build_plan.rs:56:5
   |
56 |     pub fn or(mut self) -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn or(mut self) -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/data/build_plan.rs:65:5
   |
65 |     pub fn build(self) -> BuildPlan {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn build(self) -> BuildPlan`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
   --> src/data/buildpack.rs:180:5
    |
180 |     pub fn as_str(&self) -> &str {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn as_str(&self) -> &str`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
   --> src/data/buildpack.rs:220:5
    |
220 |     pub fn as_str(&self) -> &str {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn as_str(&self) -> &str`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this function could have a `#[must_use]` attribute
 --> src/data/defaults.rs:1:1
  |
1 | pub const fn r#false() -> bool {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub const fn r#false() -> bool`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/data/launch.rs:33:5
   |
33 |     pub fn new() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn new() -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/data/launch.rs:42:5
   |
42 |     pub fn process(mut self, process: Process) -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn process(mut self, process: Process) -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
   --> src/data/launch.rs:106:5
    |
106 |     pub fn as_str(&self) -> &str {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn as_str(&self) -> &str`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
   --> src/layer_env.rs:116:5
    |
116 |     pub fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn new() -> Self`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
   --> src/layer_env.rs:146:5
    |
146 |     pub fn apply(&self, target: TargetLifecycle, env: &Env) -> Env {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn apply(&self, target: TargetLifecycle, env: &Env) -> Env`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/env.rs:41:5
   |
41 |     pub fn from_current() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn from_current() -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/env.rs:46:5
   |
46 |     pub fn new() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn new() -> Self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate

warning: this method could have a `#[must_use]` attribute
  --> src/env.rs:69:5
   |
69 |     pub fn iter(&self) -> std::collections::hash_map::Iter<'_, OsString, OsString> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn iter(&self) -> std::collections::hash_map::Iter<'_, OsString, OsString>`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate
```

GUS-W-10222390.

### PR DESCRIPTION
**Pull request by lillianzhang331**
https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate
Fixes #57

### CODE REVIEW

**Review by Malax — APPROVED** [review]

**Review by Malax — CHANGES_REQUESTED** [review]
After discussing this PR with @edmorley, I remove my approve. Adding `must_use` makes sense, but we need to go further than just slapping annotations on anything where static analysis can figure out the value needs to be used.

We most likely want to add messages to most of them too.

File: libcnb/src/build.rs [hunk 1]
```diff
@@ -72,18 +73,21 @@ impl BuildResultBuilder {
...
 impl BuildResultBuilder {
+    #[must_use]
     pub fn build(self) -> BuildResult {
         BuildResult(InnerBuildResult::Pass {
             launch: self.launch,
             store: self.store,
         })
     }
 
+    #[must_use]
```
Comment (edmorley) [inline libcnb/src/build.rs:84]: I don't think `#[must_use]` should be used on this function and the one below given they do mutate the original struct - seems like false positives.

Fixing this lint is definitely one where we need to vet each case and not just add the annotations where clippy suggested without checking :-)

**Comment by edmorley** [comment]
I found some guidelines on when to use `must_use`, here:
https://github.com/rust-lang/rust/issues/48926#issuecomment-769286972

**Comment by Malax** [comment]
Should we close this PR and continue discussion in the issue itself, or create a new one? This PR will not be merged anytime soon and with a very different changeset, correct?

**Comment by edmorley** [comment]
I think we should close this for now, and discuss in the issue what strategy we wish to use here before opening a new PR.

**Pull request closed without merging**